deps/build.jl

```diff
@@ -1,28 +1,146 @@
 using CBindingGen
+using Libdl
+import Pkg.TOML
 import P4est_jll
 
-# Get include directory from jll package
-const incdir = joinpath(dirname(dirname(P4est_jll.libp4est_path)), "include")
+
+# setup configuration using ideas from MPI.jl
+config_toml = joinpath(first(DEPOT_PATH), "prefs", "P4est.toml")
+mkpath(dirname(config_toml))
+
+if !isfile(config_toml)
+  touch(config_toml)
+end
+
+config = TOML.parsefile(config_toml)
+
+# P4est.toml has 4 keys
+#  binary   = "" (default) | "system"
+#  path     = "" (default) | top-level directory location
+#  library  = "" (default) | library name/path | list of library names/paths
+#  include  = "" (default) | include name/path | list of include names/paths
+
+
+# Step 1: Check environment variables and update preferences accordingly
+if haskey(ENV, "JULIA_P4EST_BINARY")
+	config["binary"] = ENV["JULIA_P4EST_BINARY"]
+elseif haskey(ENV, "JULIA_P4EST_LIBRARY") || haskey(ENV, "JULIA_P4EST_PATH")
+	config["binary"] = "system"
+else
+	config["binary"] = ""
+end
+
+if haskey(ENV, "JULIA_P4EST_PATH")
+	config["path"] = ENV["JULIA_P4EST_PATH"]
+else
+	config["path"] = ""
+end
+
+if haskey(ENV, "JULIA_P4EST_LIBRARY")
+	config["library"] = ENV["JULIA_P4EST_LIBRARY"]
+else
+	config["library"] = ""
+end
+
+if haskey(ENV, "JULIA_P4EST_INCLUDE")
+	config["include"] = ENV["JULIA_P4EST_INCLUDE"]
+else
+	config["include"] = ""
+end
+
+if haskey(ENV, "JULIA_P4EST_MPI_INCLUDE")
+	config["mpi_include"] = ENV["JULIA_P4EST_MPI_INCLUDE"]
+else
+	config["mpi_include"] = ""
+end
+
+
+open(config_toml, "w") do io
+	TOML.print(io, config)
+end
+
+
+
+# Step 2: Choose the library according to the settings
+if config["binary"] == ""
+	# TODO
+elseif config["binary"] == "system"
+	# TODO
+end
```

Comment:
After giving it some thought, I am not sure if allowing this `JULIA_P4EST_BINARY` envvar even makes sense. In contrast to MPI, we are not just looking for a shared library binary but also for the corresponding header files. Since unlike `find_library`, there is no `find_include` function (unless https://discourse.julialang.org/t/is-there-a-find-include-find-header-function-similar-to-libdl-s-find-library-function/48145 turns up something, which I doubt).

Thus instead of `JULIA_P4EST_BINARY`, our entry point should be `JULIA_P4EST_PATH`: If empty, use the default values from `P4est_jll.jl`, otherwise start the search from there (using `$JULIA_P4EST_PATH/include` and `$JULIA_P4EST_PATH/lib` as the default search paths for headers & library). What do you think?

Reply:
Yeah, that's an artifact from copying over some code from MPI.jl. I've reduced the possible settings to `JULIA_P4EST_LIBRARY` and `JULIA_P4EST_INCLUDE`.

Reply:
We could also provide an argument `JULIA_P4EST_PATH`, but that's not necessarily helpful. For example, it doesn't help with default installation on Debian from the official package repositories. Hence, I think it's okay to ask the user explicitly for the information that we want to have.